modules/images/image-loading-optimization/class-ilo-url-metrics-group.php

```diff
@@ -0,0 +1,172 @@
+		$this->minimum_viewport_width = $minimum_viewport_width;
+		$this->maximum_viewport_width = $maximum_viewport_width;
+		$this->sample_size            = $sample_size;
+		$this->freshness_ttl          = $freshness_ttl;
+	}
+
+	/**
+	 * Gets the minimum possible viewport width (inclusive).
+	 *
+	 * @return int Minimum viewport width.
+	 */
+	public function get_minimum_viewport_width(): int {
+		return $this->minimum_viewport_width;
+	}
+
+	/**
+	 * Gets the maximum possible viewport width (inclusive).
+	 *
+	 * @return int Minimum viewport width.
+	 */
+	public function get_maximum_viewport_width(): int {
+		return $this->maximum_viewport_width;
+	}
+
+	/**
+	 * Checks whether the provided viewport width is within the minimum/maximum range for
+	 *
+	 * @param int $viewport_width Viewport width.
+	 * @return bool Whether the viewport width is in range.
+	 */
+	public function is_viewport_width_in_range( int $viewport_width ): bool {
+		return (
+			$viewport_width >= $this->minimum_viewport_width &&
+			$viewport_width <= $this->maximum_viewport_width
+		);
+	}
+
+	/**
+	 * Adds a URL metric to the group if it is in the bounds of the max and min viewport width.
+	 *
+	 * @param ILO_URL_Metric $url_metric URL metric.
+	 * @return bool Whether the URL metric was added.
+	 */
+	public function add_url_metric( ILO_URL_Metric $url_metric ): bool {
+		if ( ! $this->is_viewport_width_in_range( $url_metric->get_viewport()['width'] ) ) {
+			return false;
+		}
+
+		$this->url_metrics[] = $url_metric;
+
+		// If we have too many URL metrics now, remove the oldest ones up to the sample size.
+		if ( count( $this->url_metrics ) > $this->sample_size ) {
+
+			// Sort URL metrics in descending order by timestamp.
+			usort(
+				$this->url_metrics,
+				static function ( ILO_URL_Metric $a, ILO_URL_Metric $b ): int {
+					return $b->get_timestamp() <=> $a->get_timestamp();
+				}
+			);
+
+			// Only keep the sample size of the newest URL metrics.
+			$this->url_metrics = array_slice( $this->url_metrics, 0, $this->sample_size );
+		}
+		return true;
+	}
+
+	/**
+	 * Determines whether the URL metrics group is complete.
+	 *
+	 * A group is complete if it has the full sample size of URL metrics
+	 * and all of these URL metrics are fresh.
+	 *
+	 * @return bool Whether complete.
+	 */
+	public function is_complete(): bool {
+		if ( count( $this->url_metrics ) < $this->sample_size ) {
+			return false;
+		}
+		$current_time = microtime( true );
+		foreach ( $this->url_metrics as $url_metric ) {
+			if ( $current_time > $url_metric->get_timestamp() + $this->freshness_ttl ) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Gets the URL metrics in the group.
+	 *
+	 * @return ILO_URL_Metric[] URL metrics in group.
+	 */
+	public function get_url_metrics(): array {
+		return $this->url_metrics;
+	}
+
+	/**
+	 * Counts the URL metrics in the group.
+	 *
+	 * @return int Count of URL metrics in the group.
+	 */
+	public function count(): int {
```

Comment:
I recall there being some recommendation not to use certain method names for "reserved" PHP terms like `count`. I can't pinpoint it, but in any case it may be safer to go with another name. How about `get_count()` or `get_size()` or just `size()`?

Alternatively, you could implement the `Countable` interface to be able to call `count( $url_metrics_group )` on the class instance itself. Not a must have, but worth considering, alongside other array interfaces like `Traversable`, which would allow doing `foreach ( $url_metrics_group as $url_metric )` on the class instance.

Reply:
I implemented `Countable` and `IteratorAggregate` in 0cce1326370817c2be9841c9a0a02213050dd3f7 and ebb74327169b0b0407b49fb729ee02c8342f2119.